File: EduLink.Domain/Entities/Cliente.cs
using EduLink.Domain.Enums;
using EduLink.Domain.Interfaces;

namespace EduLink.Domain.Entities;

public class Cliente : Usuario, IPuedeReservar
{
    public List<Reserva> Historial { get; private set; } = new();
    public List<string> MetodosPagoTokenizados { get; private set; } = new();

    public void Reservar(Servicio servicio, SlotHorario slot)
    {
        if (slot.Estado != EstadoSlot.Disponible)
            throw new InvalidOperationException("El slot no está disponible.");

        var nuevaReserva = new Reserva(this, servicio, slot);
        Historial.Add(nuevaReserva);
        // Notificación se disparará luego con Observer, creo
    }
}

File: app/domain/entities/cliente.rb
require_relative './usuario'
require_relative '../interfaces/i_puede_reservar'
require_relative '../enums/estado_slot'
require_relative './reserva'

# Forward declarations
class Servicio; end
class SlotHorario; end

module Entities
  class Cliente < Usuario
    include Interfaces::IPuedeReservar # Declara que cumple con el contrato

    attr_accessor :historial, :metodos_pago_tokenizados

    def initialize(attributes = {})
      super
      @historial ||= []
      @metodos_pago_tokenizados ||= []
    end

    # Implementación del método 'reservar' como exige el contrato
    def reservar(servicio, slot)

      nueva_reserva = Entities::Reserva.new(self, servicio, slot)
      self.historial << nueva_reserva
      
      nueva_reserva
    end
  end
end
